src/native/fs/fs_napi.cpp

```diff
@@ -0,0 +1,747 @@
+#include <vector>
+// #include <errno.h>
+// #include <sys/fsuid.h>
+
+namespace noobaa
+{
+
+struct Entry {
+    std::string name;
+    ino_t ino;
+    uint8_t type;
+};
+
+template <typename T>
+static Napi::Value api(const Napi::CallbackInfo& info)
+{
+    auto w = new T(info);
+    Napi::Promise promise = w->_deferred.Promise();
+    w->Queue();
+    return promise;
+}
+
+/**
+ * FSWorker is a general async worker for our fs operations
+ */
+struct FSWorker : public Napi::AsyncWorker
+{
+    Napi::Promise::Deferred _deferred;
+    int _errno;
+    std::string _desc;
+
+    FSWorker(const Napi::CallbackInfo& info)
+        : AsyncWorker(info.Env())
+        , _deferred(Napi::Promise::Deferred::New(info.Env()))
+        , _errno(0)
+    {
+    }
+    void Begin(std::string desc)
+    {
+        _desc = desc;
+        LOG("FS::FSWorker::Begin: " << _desc);
+    }
+    void SetSyscallError()
+    {
+        _errno = errno;
+        std::string errmsg = strerror(errno);
+        SetError(errmsg);
+    }
+    virtual void OnOK()
+    {
+        LOG("FS::FSWorker::OnOK: undefined " << _desc);
+        _deferred.Resolve(Env().Undefined());
+    }
+    virtual void OnError(Napi::Error const &error)
+    {
+        Napi::Env env = Env();
+        LOG("FS::FSWorker::OnError: " << _desc << " " << DVAL(error.Message()));
+        auto obj = error.Value();
+        obj.Set("code", Napi::String::New(env, uv_err_name(uv_translate_sys_error(_errno))));
+        _deferred.Reject(obj);
+    }
+};
+
+/**
+ * Stat is an fs op
+ */
+struct Stat : public FSWorker
+{
+    std::string _path;
+    struct stat _stat_res;
+
+    Stat(const Napi::CallbackInfo& info) : FSWorker(info)
+    {
+        _path = info[0].As<Napi::String>();
+        Begin(XSTR() << DVAL(_path));
+    }
+    virtual void Execute()
+    {
+        int r = stat(_path.c_str(), &_stat_res);
+        if (r != 0) SetSyscallError();
+    }
+    virtual void OnOK()
+    {
+        LOG("FS::Stat::OnOK: " << DVAL(_path) << DVAL(_stat_res.st_ino) << DVAL(_stat_res.st_size));
+        Napi::Env env = Env();
+        auto res = Napi::Object::New(env);
+        res["dev"] = Napi::Number::New(env, _stat_res.st_dev);
+        res["ino"] = Napi::Number::New(env, _stat_res.st_ino);
+        res["mode"] = Napi::Number::New(env, _stat_res.st_mode);
+        res["nlink"] = Napi::Number::New(env, _stat_res.st_nlink);
+        res["uid"] = Napi::Number::New(env, _stat_res.st_uid);
+        res["gid"] = Napi::Number::New(env, _stat_res.st_gid);
+        res["rdev"] = Napi::Number::New(env, _stat_res.st_rdev);
+        res["size"] = Napi::Number::New(env, _stat_res.st_size);
+        res["blksize"] = Napi::Number::New(env, _stat_res.st_blksize);
+        res["blocks"] = Napi::Number::New(env, _stat_res.st_blocks);
+        res["atime"] = Napi::Number::New(env, _stat_res.st_atime);
+        res["mtime"] = Napi::Number::New(env, _stat_res.st_mtime);
+        res["ctime"] = Napi::Number::New(env, _stat_res.st_ctime);
```

Comment:
Based on what we discussed below on times - we need to convert the times to milliseconds to be compatible with js dates. Also need to take into account the nano seconds precision which requires some difference between Mac and Linux. This example works on both linux and mac:

```
$ cat a.cpp
#include <stdio.h>
#include <sys/stat.h>
int main()
{
	struct stat st;
	int r = stat(".", &st);
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
	double d = (double(1e3) * st.st_mtimespec.tv_sec) + (double(1e-6) * st.st_mtimespec.tv_nsec);
#else
	double d = (double(1e3) * st.st_mtim.tv_sec) + (double(1e-6) * st.st_mtim.tv_nsec);
#endif
	printf("mtime = %lf \n", d);
	return 0;
}
```